### ISSUE TITLE
fit bortfeld function with adai and adam, adai cannot converge at the same learning rate

### ISSUE THREAD

**Issue by civerjia** [zeke-xie/adaptive-inertia-adai#4]
I tested some simple models on pytorch, Adai does have better performance over Adam, therefore, I tried to use Adai to fit bortfeld function. I implemented two matlab functions to compare the performance of Adai and Adam.  And I found Adai cannot converge while Adam converged. Is my implementation wrong?

Bortfeld function() is a function used to fit proton bragg peak, [An analytical approximation of the Bragg curve for therapeutic proton beams](https://aapm.onlinelibrary.wiley.com/doi/abs/10.1118/1.598116)

```
function [theta_best,loss] = adai(depth,para,idd_i,lb,ub,lr)
    % adam inertia
    T = 2000;
    beta0 = 0.1;
    beta1_cum_prod = 1;
    beta2 = 0.99;
    epsilon = 1e-3;
    loss = zeros(T,1);
    m_tm1 = 0;
    v_tm1 = 0;
    theta_tm1 = para;
    v_t_mean = 0;
    
    theta_best = para;
    loss_best = 1e9;
    loss(1) = norm((bf_mex(depth,theta_tm1,'idd') - idd_i),'fro');
    for t = 2:T
        % get gradient = jacobian*error
        g_t = 2*bf_mex(depth,theta_tm1,'jacobian')'*(bf_mex(depth,theta_tm1,'idd') - idd_i);
        % Update biased second raw moment estimate
        v_t = beta2*v_tm1 + (1-beta2)*g_t.^2;
        % Compute bias-corrected second raw moment estimate
        v_t_hat = v_t / (1-beta2^(t-1));
        v_t_mean = mean(v_t_hat);
        beta1t = max(min(1-(v_t_hat./v_t_mean).*beta0, 1-epsilon),0);
        % Update biased first moment estimate
        m_t = beta1t.*m_tm1 + (1-beta1t).*g_t;
        beta1_cum_prod = beta1_cum_prod.*beta1t;
        % Compute bias-corrected first moment estimate
        m_t_hat = m_t ./ (1-beta1_cum_prod);
        % Update parameters
        theta_t = theta_tm1 - lr*m_t_hat;
        % constrain
        theta_t(theta_t < lb) = lb(theta_t < lb);
        theta_t(theta_t > ub) = ub(theta_t > ub);
        
        theta_tm1 = theta_t;
        m_tm1 = m_t;
        v_tm1 = v_t;
            
        idd_pred = bf_mex(depth,theta_t,'idd');
        loss(t) = norm((idd_pred - idd_i),'fro');
        
        if loss(t) < loss_best
           loss_best = loss(t);
           theta_best = theta_t;
        end
        if (abs(loss(t) - loss(t-1)) < 1e-6)
            break;
        end
    end
    
end
```

```
function [theta_best,loss] = adam(depth,para,idd_i,lb,ub,lr)
    T = 2000;
    beta1 = 0.9;
    beta2 = 0.999;
    epsilon = 1e-8;
    loss = zeros(T,1);
    m_tm1 = 0;
    v_tm1 = 0;
    theta_tm1 = para;
    
    theta_best = para;
    loss_best = 1e9;
    loss(1) = norm((bf_mex(depth,theta_tm1,'idd') - idd_i),'fro');
    for t = 2:T
        % get gradient = jacobian*error
        g_t = 2*bf_mex(depth,theta_tm1,'jacobian')'*(bf_mex(depth,theta_tm1,'idd') - idd_i);
        % Update biased first moment estimate
        m_t = beta1*m_tm1 + (1-beta1)*g_t;
        % Update biased second raw moment estimate
        v_t = beta2*v_tm1 + (1-beta2)*g_t.^2;
        % Compute bias-corrected first moment estimate
        m_t_hat = m_t / (1-beta1^(t-1));
        % Compute bias-corrected second raw moment estimate
        v_t_hat = v_t / (1-beta2^(t-1));
        % Update parameters
        theta_t = theta_tm1 - lr*m_t_hat./(sqrt(v_t_hat)+epsilon);
        
        % constrain
        theta_t(theta_t < lb) = lb(theta_t < lb);
        theta_t(theta_t > ub) = ub(theta_t > ub);
        
        theta_tm1 = theta_t;
        m_tm1 = m_t;
        v_tm1 = v_t;
            
        idd_pred = bf_mex(depth,theta_t,'idd');
        loss(t) = norm((idd_pred - idd_i),'fro');
        
        if loss(t) < loss_best
           loss_best = loss(t);
           theta_best = theta_t;
        end
        if (abs(loss(t) - loss(t-1)) < 1e-6)
            break;
        end
    end
    
end
```

**Comment by civerjia** [comment]
For your convenient, I create a repo : https://github.com/civerjia/bortfeld_function_adai_adam.git

**Comment by zeke-xie** [comment]
Hi civerjia,

I am not familiar with Matlab programming. 

I have two questions about the experimental details.

Could you please tell me the hyperparameters (e.g. learning rate, learning rate scheduler) of Adai and Adam which you used in your experiments?

When you say "cannot converge", do you mean 1) the training loss becomes infinite or 2) the training loss is not stable in final epochs or 3) something else? Could you specify it more clearly?

These details are important for understanding non-convergence, theoretically and empirically.

**Comment by civerjia** [comment]
learning rate is constant, I set the lr = 1e-2,1e-3,1e-4 for 3 experiments
hyperparameter :
Adai  beta0 = 0.1, beta2 = 0..99, epsilon = 1e-3
Adam beta1 = 0.9, beta2 = 0.999, epsilon = 1e-8

When lr = 1e-4, Adai works but it's too slow compared with Adam@lr=1e-2

**Comment by civerjia** [comment]
Loss:
![0 01](https://user-images.githubusercontent.com/29112974/180612764-277a372d-3768-41dd-96e5-2dee509d86dd.png)
![0 001](https://user-images.githubusercontent.com/29112974/180612767-c203695e-1c69-4032-a515-2a3fd805fc8f.png)
![0 0001](https://user-images.githubusercontent.com/29112974/180612769-ea01e75c-39f2-4340-8398-555b0dbf0923.png)

**Comment by zeke-xie** [comment]
Hi civerjia,

I am able to review the Matlab code in your experiments, as I am not very familiar with Matlab programming.

Given the experimental result you presented, I would not say it is a problem of "cannot converge".

Instead, you actually mean that Adam seems to converge faster than Adai given the specified loss and the specified hyperparameters. 

I personally think this is not a problem.
For a specified optimization setting without tuning hyperparameters (e.g. lr scheduler), it is common to see that one optimizer converges faster than another.

I am not an expert on your problem setting.

I shared some hyperparameters strategies in readme. 
They may be helpful for you to improve the optimization performance of Adai in your task.
BTW, the learning rate scheduler is very important for good and fast convergence.

**Comment by civerjia** [comment]
Hyperparameters tuning is very important, but it doesn't work in this case, I tried beta0 = 0:0.1:0.9 and beta2 = 0.1:0.05:0.99.  And I got the similar results, If I start with a large learning rate and decrease it during the training, it won't work. Adai got trapped somewhere, the loss is constant, and Adai cannot escape from it. Even a cosine annealing schedule can not restart it.
And I checked the code, it's caused by the upper bound and lower bound, it might go to inf without these bounds(why I said not converged).

**Comment by zeke-xie** [comment]
Please rethink the hyperparameters strategies in readme. 
Adai and SGD may share similar optimal settings.

As I am not an expert on your problem setting, I have no more suggestion on what is a good scheduler in this case. 

If you have no more question about Adai itself, I will close the issue.

**Comment by civerjia** [comment]
> 

Final update: I build a PyTorch version if you want to try it. https://github.com/civerjia/bortfeld_function_adai_adam.git

![Loss_lr0 01](https://user-images.githubusercontent.com/29112974/180672416-81006597-ad9b-4e70-bcba-e41afadc3c9f.png)

**Comment by zeke-xie** [comment]
Your training curves shows that the initial learning rate is too large for SGD and Adai.

As I mentioned above, designing the learning rate scheduler is a separated problem and highly problem-dependent.

So I will close this issue.
